plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TypeConverter.java

```diff
@@ -179,25 +192,41 @@ private static org.apache.iceberg.types.Type fromDecimal(DecimalType type)
         return Types.DecimalType.of(type.getPrecision(), type.getScale());
     }
 
-    private static org.apache.iceberg.types.Type fromRow(RowType type)
+    private static org.apache.iceberg.types.Type fromRow(RowType type, Optional<ColumnIdentity> columnIdentity, Optional<AtomicInteger> nextFieldId)
     {
         List<Types.NestedField> fields = new ArrayList<>();
-        for (RowType.Field field : type.getFields()) {
+
+        for (int i = 0; i < type.getFields().size(); i++) {
+            int index = i;
+            Optional<ColumnIdentity> childColumnIdentity = columnIdentity.map(column -> column.getChildren().get(index));
+            int id = childColumnIdentity.map(ColumnIdentity::getId).orElseGet(() -> nextFieldId.orElseThrow(() ->
+                    new TrinoException(INVALID_PARAMETER_USAGE, "Create new column need nextFieldId.")).getAndIncrement());
```

Comment:
A more accurate exception would be:

```
new IllegalArgumentException("Either a column identity or nextFieldId is expected"); 
```

IMO we shouldn't throw this as a user error - this is more like a validation